certbot/docs/contributing.rst

```diff
@@ -338,17 +338,57 @@ it was not installed properly.
 Once you've finished your plugin and published it, you can have your
 users install it system-wide with `pip install`. Note that this will
 only work for users who have Certbot installed from OS packages or via
-pip. Users who run `certbot-auto` are currently unable to use third-party
-plugins. It's technically possible to install third-party plugins into
-the virtualenv used by `certbot-auto`, but they will be wiped away when
-`certbot-auto` upgrades. If you'd like your plugin to be used alongside
-the Certbot snap, you will also have to publish your plugin as a snap.
-Certbot's DNS plugins and the README file in ``tools/snap/`` provide a
-starting reference for how to do this.
+pip.
 
 .. _`setuptools entry points`:
     https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entry-points
 
+Writing your own plugin snap
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you'd like your plugin to be used alongside the Certbot snap, you
+will also have to publish your plugin as a snap. Plugin snaps are
+regular confined snaps, but normally do not provide any "apps"
+themselves. Plugin snaps export loadable Python modules to the Certbot
+snap.
+
+When the Certbot snap runs, it will use its version of Python and prefer
+Python modules contained in its own snap over modules contained in
+external snaps. This means that your snap doesn't have to contain things
+like an extra copy of Python, Certbot, or their dependencies, but also
+that if you need a different version of a dependency than is already
+installed in the Certbot snap, the Certbot snap will have to be updated.
+
+Certbot plugin snaps expose their Python modules to the Certbot snap via
```

Comment:
Sigh. So, we should probably mention that we're not actually doing this properly. "Properly" would be to set things up so the file appears neatly within `/snap/certbot/`, like we do with the metadata interface. We don't actually do that; we just have our plugins install files to their python site packages folder, and certbot goes out and gets them. Or we could fix this. I didn't realize this is what we were using `jq` for that was causing problems, I just looked and I see now that when we translated to python, we still kept doing it this way. Which is fine, and works, but isn't actually idiomatic because we have a classic snap.

Reply:
This to me seems like an implementation detail of the core Certbot snap that authors of external plugins shouldn't have to worry about in the context of docs titled "Writing your own plugin snap". Am I overlooking something that makes this relevant to them?

If not, I agree that documenting for ourselves that we're doing something strange here is a good idea (even if it's just done in code comments), but I would personally prefer if it was done in a different PR and not including in the docs for writing externally snapped plugins.

What do you think?

Reply:
I imagined that a plugin author would go read the snapcraft documentation, understand how the content interface works, then become confused about where things go. If we're explicitly relying on packages being available in the plugin's site-packages folder, we should document that because normally when building a snap there's no reason to think that your own implementation detail of how you set up your python packages would be relevant. By the standard api, the only thing that *should* be relevant is the things you explicitly expose to the content interface, but since we're not using that (and essentially breaking the api), we really should be saying what the actual api adherence that we need is. It's unlikely to come up in practice, sure, because we're assuming that plugin authors follow python defaults, but we should make it explicit that we're making that assumption.

Reply:
Oh I see! The piece I wasn't understanding was our reliance on things being in `lib/python3.8/site-packages/`. Thanks for clarifying and pushing back. I'll document this.